### PR TITLE
rm unnecessary echos

### PR DESCRIPTION
**Pull request by wilhelmtell**
similar to avoiding a cat pipe with a redirection, one can avoid an echo
pipe with a here string.

### CODE REVIEW

**Comment by waldyrious** [comment]
IMHO, removing `cat` and `echo` isn't always a net benefit, since whatever overhead they add is imperceptible, and removing them (to add their arguments at the end instead) makes the code harder to read, since that entails mixing right-to-left constructs into left-to-right expressions.

**Comment by wilhelmtell** [comment]
> removing them (to add their arguments at the end instead) makes the code harder to read, since that entails mixing right-to-left constructs into left-to-right expressions.

How is right-to-left harder to read than left-to-right? How is mixing the two hard to read? Surely you don't avoid postfix increments too. Surely you don't avoid mixing pipes and command line arguments. This is not hard to read. It's idiomatic, it's coming straight from the man page. And it gets the same thing said with less code. That is net good.

**Comment by nabijaczleweli** [comment]
> How is right-to-left harder to read than left-to-right? 

Because data _always_ flows-like-this-> not <-this-like-flows, maybe?

File: with [hunk 1]
```diff
@@ -20,20 +20,20 @@ while [ true ]; do
...
       # remove with
       if [ $curr_command == - ]; then
-        prefix=$( echo $prefix | awk '{$NF="";sub(/[ \t]+$/,"")}1' )
+        prefix=$( awk '{$NF="";sub(/[ \t]+$/,"")}1' <<<"$prefix" )
         pmpt="$prefix>"
       # nesting withs
       elif [[ $curr_command == +* ]]; then
         parsed=${options#"+"}
-        prefix=$( echo "$prefix" "$parsed" )
+        prefix="$prefix $parsed"
```
Comment (mchav) [inline with:36]: The problem with this is that it will expand wild cards.

File: with [hunk 2]
```diff
@@ -20,20 +20,20 @@ while [ true ]; do
   while IFS="" read -r -e -d $'\n' -p "$pmpt " options; do
     history -s "$options"
     
-    curr_command="$(echo $options | { read first rest ; echo $first ; })" 
+    curr_command="$({ read first rest ; echo $first ; } <<<"$options")"
```
Comment (mchav) [inline with:23]: I'm not swayed either way about the direction of input. What's wrong with echoes?

Comment (wilhelmtell) [inline with:23]: In the spirit of Occam's razor, that they're not necessary. You can say any of

```
process <<<a
echo a |process
echo a |cat |process
echo a |cat |tac |cat |process
```

or any of the other infinite possible ways out there to get the same task done. How do you choose which way to go about this? What's wrong with echo? Nothing, you can use echo. What's wrong with cat? Nothing, you can use that. Or tac, wonderful utility, you can use that too. There's nothing wrong with any of them. They're just not necessary here, any of them. And that's what's wrong with this. Doing something that's not necessary. Writing more code than you must.

Okay, I've clarified my point I think. If this is an opinionated matter (can I say in my opinion it isn't?) then you can close this.

**Pull request closed without merging**